docs: add localization design doc (#141)
autowarefoundation/autoware-documentation: docs/design/autoware-architecture/localization/index.md

Developers can optionally add other frames such as odom or base_footprint as long as the tf structure above is maintained.

### The localization module's ideal functionality

- The localization module should provide pose, velocity, and acceleration for control, planning, and perception.
- Latency and stagger should be sufficiently small or adjustable such that the estimated values can be used for control within the ODD (Operational Design Domain).
- The localization module should produce the pose on a fixed coordinate frame.
- Sensors should be independent of each other so that they can be easily replaced.
- The localization module should provide a status indicating whether or not the autonomous vehicle can operate with the self-contained function or map information.
- Tools or manuals should describe how to set proper parameters for the localization module
- Valid calibration parameters should be provided to align different frame or pose coordinates and sensor timestamps.

### KPI

To maintain sufficient pose estimation performance for safe operation, the following metrics are considered:

- Safety

>   - The distance travelled within the ODD where pose estimation met the required accuracy, divided by the overall distance travelled within the ODD, as a percentage.

Reviewer: ```suggestion
  - The distance traveled within the ODD where pose estimation met the required accuracy, divided by the overall distance traveled within the ODD, as a percentage.
```

LOCALIZATION COMPONENT DESIGN DOC

## Abstract

## 1. Requirements

Localization aims to estimate vehicle pose, velocity, and acceleration.

Goals:

> - Proposition of a system that can estimate the vehicle pose, velocity, and acceleration as long as possible
> - Proposition of a system that can diagnose the stability of estimation and send a warning message to the error-monitoring system if the estimation result is unreliable
> - Vehicle localization function can adequately work with various sensor configurations

Reviewer: I am not 100% confident about this suggestion. I do note that the original phrasing of the third sentence does not fit with the first two goals, so some additional thinking may be required. 

```suggestion
- Propose a system that can estimate vehicle pose, velocity, and acceleration for as long as possible
- Propose a system that can diagnose the stability of estimation and send a warning message to the error-monitoring system if the estimation result is unreliable
- Design a vehicle localization function can adequately work with various sensor configurations
```

Reviewer: Oops. Thank you for that. I have updated the suggestion accordingly.

Reviewer: I also removed the word "adequately" from the third point, since it has a negative feeling to it and I do not think it is necessary to include in the sentence.

Non-goals:

> - We do not try to develop a localization system that
>   - does not fail in any environment
>   - works outside of the pre-defined ODD (Operational Design Domain)
>   - has unnecessarily better performance than required in autonomous driving

Reviewer: ```suggestion
- This design document does not aim to develop a localization system that
  - does not fail in any environment
  - works outside of the pre-defined ODD (Operational Design Domain)
  - has unnecessarily better performance than is required for autonomous driving
```

Reviewer: It was more about trying to avoid a double negative ("...does not aim to.... does not fail") but if it changes the meaning then I can revert that first bullet point back to its original.

Author: OK. I think your suggestion `is infallible in all environments` is better than the original.

## 2. Sensor Configuration Examples

> In this section, we show example sensor configurations and their expected
> performances. Each sensor has advantages and disadvantages. Therefore, by fusing multiple sensors, we can improve the overall performance.

Reviewer: ```suggestion
This section shows example sensor configurations and their expected performances. 
Each sensor has its own advantages and disadvantages, but overall performance can be improved by fusing multiple sensors.
```

- The environment has laser beams of the same frequency as LiDAR

#### Functionalities

- The system can estimate the vehicle location on the point cloud map with the error of ~10cm
- The system is Operable at the night

### 3D-LiDAR or Camera + Vector Map

#### Expected situation

- Road with clear white lines and loose curvatures, such as highway or ordinary local road

#### Situation that can make the system unstable

- White lines are scratchy or covered by rain or snow
- Tight curvature such as intersections
- Large reflection change of the road surface caused by rain or paint

#### Functionalities

- Correct vehicle positions along the lateral direction
- Pose correction along the longitudinal can be inaccurate. But we can resolve this by fusing with GNSS

### GNSS

#### Expected situation

- The vehicle is placed in an open environment with no or few surrounding objects, such as a rural landscape

#### Situation that can make the system unstable

- GNSS signals are intercepted by surrounding objects, e.g., tunnels or buildings

#### Functionalities

- The system can estimate the vehicle position with an error of ~10m in the world coordinate

> - With RTK attached, the accuracy can be improved to ~10cm

Reviewer: @IshitaTakeshi Could you answer the question above please?

> - With RTK attached, the accuracy can be improved to ~10cm

Reviewer 2: Sorry, this word is too abbreviated.
RTK -> RTK-GNSS -> Real Time Kinematic Global Navigation Satellite System

> - The system with this configuration can work without environment maps

Reviewer 3: yes. It means both

> - The system with this configuration can work without environment maps

Reviewer: Suggestion updated!

### Camera (Visual Odometry, Visual SLAM)

#### Expected situation

> - The vehicle is placed in a texture-rich environment, such as urban areas

Reviewer: Suggestion updated!

#### Situation that can make the system unstable

- The vehicle is placed in a Texture-less environment
- The vehicle is surrounded by other objects
- The camera observes significant illumination changes such as sunshine, headlight, or brightness change at the exit of a tunnel
- The vehicle is placed in a dark environment

#### Functionalities

- The system can estimate the odometry by tracking visual features

### Wheel speed sensor

#### Expected situation

> - The vehicle is running on an ordinary road

Author: I think this implies that the wheel does not slip on the road, so if I replace this I would say "a flat and smooth road" @YamatoAndo @e-takeuchi

> - The vehicle is running on an ordinary road

Reviewer 3: OK. I agree with @IshitaTakeshi san.

> - The vehicle is running on an ordinary road

Reviewer: Suggestion updated!

> - The vehicle is running on an ordinary road

Reviewer 3: OK!

#### Situations that can make the system unstable

- The vehicle is running on a slippery or bumpy road, which can cause incorrect observations of wheel speed.

#### Functionality

- The system can acquire the vehicle velocity and estimate distance traveled.

### IMU

#### Expected environments

> - Ordinary roads

Reviewer: ```suggestion
- Flat, smooth roads
```

> <!-- NOTE: The availability of IMU is highest among the sensors attached to vehicles. Therefore we should list more situations or say we can use the IMU almost anywhere -->

Reviewer: ```suggestion
```

Author: Resolved in c8031864

- Ordinary roads

<!-- NOTE: The availability of IMU is highest among the sensors attached to vehicles. Therefore we should list more situations or say we can use the IMU almost anywhere -->

#### Situation that can make the system unstable

> - IMUs have a bias that is dependent on the surrounding temperature. This can cause incorrect sensor observation or odometry drift

Author: The bias is a little bit complicated. The detailed explanation is written in the BIAS section of this page and especially what we need to care about the most is the turn-on bias.

https://www.vectornav.com/resources/inertial-navigation-primer/specifications--and--error-budgets/specs-imuspecs

Usually the bias gradually changes as the surrounding temperature changes. I think this is because IMU is a semiconductor, which changes its characteristics by the surrounding temperature, like a transistor.

> - IMUs have a bias that is dependent on the surrounding temperature. This can cause incorrect sensor observation or odometry drift

Reviewer: Would it be an idea to add that link to the page above if anyone wants to get a better idea about bias?

> - IMUs have a bias that is dependent on the surrounding temperature. This can cause incorrect sensor observation or odometry drift

Reviewer: Alternatively, perhaps we need to consider adding footnotes to the bottom of design docs, so that relevant information can be included but in such a way to avoid affecting the flow of the main content.

> - IMUs have a bias that is dependent on the surrounding temperature. This can cause incorrect sensor observation or odometry drift

Author: @kenji-miyake Is it ok to include an external link to the design doc page? This seems a link to a device production company

> - IMUs have a bias that is dependent on the surrounding temperature. This can cause incorrect sensor observation or odometry drift

Author: @LalithVipulananthan Yes! Footnote is a great idea!

> - IMUs have a bias that is dependent on the surrounding temperature. This can cause incorrect sensor observation or odometry drift

Reviewer 4: I think it's okay as it seems to be a link to general information.

> - IMUs have a bias that is dependent on the surrounding temperature. This can cause incorrect sensor observation or odometry drift

Reviewer 4: @LalithVipulananthan Mm? It's already enabled.
https://github.com/autowarefoundation/autoware-documentation/blob/533f0fc640541d468039d3d0d5796dec291925e8/mkdocs.yaml#L66

> - IMUs have a bias that is dependent on the surrounding temperature. This can cause incorrect sensor observation or odometry drift

Reviewer: Oops. 😅 

@IshitaTakeshi I have updated my suggestion above to add the information about IMU bias as a footnote.

#### Functionalities

- The system can observe acceleration and angular velocity.
- By integrating these observations, the system can estimate the local pose change and realize dead-reckoning

> - By fusing with external sensors, it can estimate more accurate poses in a higher frequency

Reviewer: @IshitaTakeshi Is dead-reckoning understood well enough to be included here without any explanation?

- The vehicle is placed in an environment with low magnetic noise

#### Situations that can make the system unstable

- The vehicle is placed in an environment with high magnetic noise, such as one containing buildings or structures with reinforced steel or other materials that generate electromagnetic waves.

#### Functionality

- The system can estimate the vehicle's direction in the world coordinate system.

### Magnetic markers

#### Expected situation

- The car is placed in an environment with magnetic markers installed.

#### Situations where the system becomes unstable

- The markers are not maintained.

#### Functionality

- Vehicle location can be obtained on the world coordinate by detecting the magnetic markers.
- The system can work even if the road is covered with snow.

## 3. Requirements

- By implementing different modules, various sensor configurations and algorithms can be used.
- The localization system can start pose estimation from an ambiguous initial location.
- The system can produce a reliable initial location estimation.

> - The system can manage the state of the initial location estimation (uninitialized, initializable, or non-initializable) and can report to the error monitor.

Reviewer 4: > (uninitialized, initializable, or non-initializable)

I feel it's not MECE (Mutually Exclusive and Collectively Exhaustive).

- `uninitialized` includes `initializable` and `non-initializable`.
- `initialized` is missing.

Author: So better to replace it with `(uninitialized or initialized)`?  @YamatoAndo @e-takeuchi

Reviewer 4: I guess so.

## 3. Requirements

- By implementing different modules, we accept various sensor configurations and algorithms.
- The localization system can start pose estimation from an ambiguous initial location
- The system can produce the reliability of the initial location estimation

> - The system can manage the state of the initial location estimation (uninitialized, initializable, or not) and can report to the error monitor

Author: This means non-initializable 

Resolved in c8f62180

## 4. Architecture

### Abstract

> We define two architectures: "Required" and "Recommended." Only input and output are defined in the "Required" architecture to accept various localization algorithms. Also, to improve the reusability of each module, we describe more detailed parts in the "Recommended" architecture.

Author: How about saying this? The reason we defined the recommended architecture is to improve the reusability and I think we can provide a more clear explanation by writing the reason.

```Two architectures are defined, "Required" and "Recommended", but the "Required" architecture only contains the inputs and outputs necessary to accept various localization algorithms. To improve the reusability of each module, we defined the required components of the "Recommended" architecture. A more detailed explanation is given in the "Recommended" architecture section.```

|    map    | Origin of the map coordinate (ex. MGRS origin)                                                 |
|  viewer   | User-defined frame for rviz                                                                    |
| base_link | Reference pose of the ego-vehicle (projection of the rear-axle center onto the ground surface) |
|  sensor   | Reference pose of each sensor                                                                  |

Developers can optionally add other frames such as odom or base_footprint while keeping the tf structure above.

### The localization module's ideal functionality

- The localization module should provide pose, velocity, and acceleration for control, planning, and perception
- Latency and stagger should be sufficiently small or adjustable so that the estimated values can be used for control within the ODD(Operational Design Domain)
- The localization module should produce the pose on a fixed coordinate frame
- Sensors should be independent of each other and keep the replaceability
- The localization module should provide the status indicating if the autonomous vehicle can operate or not with the self-contained function or map information.

> - Tools or manuals should describe how to set proper parameters for the localization module <!-- TODO  原文にある「適切な位置推定」とはなんだ！！！ -->

Author: Just removed the note in 888b007e

>   - The percentage of distance which the pose estimation satisfies the required accuracy per the overall distances defined in ODD

Reviewer: Suggested change based on the conversation above.

```suggestion
  - The distance travelled within the ODD where pose estimation met the required accuracy, divided by the overall distance travelled within the ODD, as a percentage.
```

Author: Resolved in e7aa1adf

>   - The accuracy of out-of-ODD detection

Reviewer: ```suggestion
  - The accuracy of detecting when the vehicle goes outside of the ODD.
```

Reviewer: > Sometimes a vehicle goes beyond the ODD. And in such a situation, we need to detect that the vehicle is operating on the outside of the ODD. So the accuracy of out-of-ODD detection means the detection rate of it.

@IshitaTakeshi I have made a suggestion based on your response.

### KPI

We jointly consider these metrics if the system keeps sufficient pose estimation performance for safe operation.

- Safety

>   - The percentage of distance which the pose estimation satisfies the required accuracy per the overall distances defined in ODD

Author: This describes the rate of 

```
The traveled distance that the pose estimator could satisfy the required accuracy (within ODD)
-------------------------------------------------------------------------------------------------------------------------------
The overall distance traveled (within ODD)
```

  - The anomaly detection rate of the case that the localization module cannot estimate the pose within ODD

>   - The accuracy of out-of-ODD detection

Author: Sometimes a vehicle goes beyond the ODD. And in such a situation, we need to detect that the vehicle is operating on the outside of the ODD. So the accuracy of out-of-ODD detection means the detection rate of it.

- Computational load
- Latency

## 5. Interface and Data Structure

## 6. Concerns, Assumptions, and Limitations

### Prerequisites of sensors and inputs

#### Sensor prerequisites

- Input data is not defective. Internal sensor observation such as IMU continuously keeps the proper frequency
- Input data has correct and exact time stamps. Estimated poses can be inaccurate or unstable if the timestamps are not exact
- Sensors are correctly mounted on the exact positions and accessible from TF. If the sensor positions are inaccurate, estimation results may be incorrect or unstable. A sensor calibration framework is required to properly obtain the sensor positions.

#### Map prerequisites

> - Pose estimation might be unstable if there is no sufficient information on the map. A testing framework is demanded to check if the map has adequate information for pose estimation.
> - Pose estimation might be unstable if the actual environment has different objects from the map. Maps need updates according to new objects and seasonal changes.
> - If multiple maps with different coordinate systems are used, the misalignment between them can affect the localization performance. Maps have to be aligned to a uniform coordinate, or we require an alignment framework.

Reviewer: The phrasing of the map prequisites is different to that for the sensor prerequisites and talks about the problems first before the prequisite, so I have reordered the sentences accordingly.

```suggestion
- Sufficient information is contained within the map.
  - Pose estimation might be unstable if there is insufficient information in the map. 
  - A testing framework is necessary to check if the map has adequate information for pose estimation.
- Map does not differ greatly from the actual environment.
  - Pose estimation might be unstable if the actual environment has different objects from the map. 
  - Maps need updates according to new objects and seasonal changes.
- Maps must be aligned to a uniform coordinate, or an alignment framework is in place. 
  - If multiple maps with different coordinate systems are used, the misalignment between them can affect the localization performance.
```

- With RTK attached, the accuracy can be improved to ~10cm
- The system with this configuration can work without environment maps

### Camera (Visual Odometry, Visual SLAM)

#### Expected situation

- The vehicle is placed in a texture-rich environment, such as urban areas

#### Situations that can make the system unstable

- The vehicle is placed in a Texture-less environment
- The vehicle is surrounded by other objects.
- The camera observes significant illumination changes, such as those caused by sunshine, headlights from other vehicles or when approaching the exit of a tunnel.
- The vehicle is placed in a dark environment.

#### Functionality

- The system can estimate odometry by tracking visual features.

### Wheel speed sensor

#### Expected situation

- The vehicle is running on an ordinary road

#### Situations that can make the system unstable

- The vehicle is running on a slippery or bumpy road, which can cause incorrect observations of wheel speed.

#### Functionality

- The system can acquire the vehicle velocity and estimate distance traveled.
- The system can estimate more accurate poses at a higher frequency by fusing with external sensors.

> <!-- NOTE: Is the second item necessary? Because it is general for any other sensors -->

Reviewer: Contents of the note need to be actioned and then the note removed.

Reviewer 3: I agree with you. Please remove the "The system can estimate ~" sentence.

### IMU

#### Expected environments

- Ordinary roads

> <!-- NOTE: The availability of IMU is highest among the sensors attached to vehicles. Therefore we should list more situations or say we can use the IMU almost anywhere -->

Reviewer: Contents of the note need to be actioned and then the note removed.

Reviewer 3: @IshitaTakeshi What is it, for example?

Author: @YamatoAndo This says that IMU can be used on ordinary roads but basically nothing prevents from using IMU in autonomous driving
If IMU odometry becomes unstable in a shaky environment, I think we should say like "A flat and smooth road"

Reviewer 3: I see. OK

#### Situations that can make the system unstable

- IMUs have a bias that is dependent on the surrounding temperature. This can cause incorrect sensor observation or odometry drift

#### Functionality

- The system can observe acceleration and angular velocity.
- By integrating these observations, the system can estimate the local pose change and realize dead-reckoning
- By fusing with external sensors, it can estimate more accurate poses in a higher frequency

> <!-- NOTE: The third item says "it can estimate more accurate poses in a higher frequency," but the IMU already has very frequent observation -->

Reviewer: Contents of the note need to be actioned and then the note removed.

Author: @YamatoAndo @e-takeuchi How about this? IMU already has a very high frequency, so I think it is not really necessary to describe the fusion with other sensors

Reviewer 3: @IshitaTakeshi I see. OK

#### Functionality

- The system can estimate the vehicle position with an error of ~10m in the world coordinate
- With RTK attached, the accuracy can be improved to ~10cm
- The system with this configuration can work without environment maps

### Camera (Visual Odometry, Visual SLAM)

#### Expected situation

- The vehicle is placed in a texture-rich environment, such as urban areas

#### Situations that can make the system unstable

- The vehicle is placed in a Texture-less environment
- The vehicle is surrounded by other objects.
- The camera observes significant illumination changes, such as those caused by sunshine, headlights from other vehicles or when approaching the exit of a tunnel.
- The vehicle is placed in a dark environment.

#### Functionality

- The system can estimate odometry by tracking visual features.

### Wheel speed sensor

#### Expected situation

- The vehicle is running on an ordinary road

#### Situations that can make the system unstable

- The vehicle is running on a slippy or bumpy road
- The vehicle cannot observe the correct wheel speed because of poor road surface conditions such as bumps or slips

#### Functionality

> - The system can acquire the vehicle velocity and estimate distance travelled.

Reviewer: ```suggestion
- The system can acquire the vehicle velocity and estimate distance traveled.
```
